java-client/src/main/java/org/apache/pegasus/client/PegasusScanner.java

```diff
@@ -88,9 +88,27 @@ public PegasusScanner(
     _needCheckHash = needCheckHash;
     _incomplete = false;
     _fullScan = fullScan;
+    _nextItem = null;
+  }
+
+  public boolean hasNext() throws PException {
+    synchronized (_promisesLock) {
+      if (_nextItem != null) {
+        return true;
+      }
+    }
+    _nextItem = next();
```

Comment:
It's not thread safe in some cases since `_nextItem` will be overwritten.

```
time  thread1         thread2
  t1  hasNext         hasNext
  t2  _nextItem = a
  t3                  _nextItem = b
  t4  next -> b
  t5                  next -> nullptr
```